java-checks/src/main/java/org/sonar/java/checks/RecordPatternInsteadOfFieldAccessCheck.java

```diff
@@ -0,0 +1,123 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.CaseLabelTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.PatternInstanceOfTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypePatternTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+
+@Rule(key = "S6878")
+public class RecordPatternInsteadOfFieldAccessCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final List<String> ALLOWED_METHODS = List.of("toString", "hashCode", "equals");
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava21Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.PATTERN_INSTANCE_OF, Tree.Kind.CASE_LABEL);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree instanceof PatternInstanceOfTree instanceOf) {
+      var pattern = instanceOf.pattern();
+      if (pattern instanceof TypePatternTree typePattern && isRecordPattern(typePattern)) {
+        checkTypePatternVariableUsage(typePattern.patternVariable());
+      }
+    } else {
+      checkCaseLabel((CaseLabelTree) tree);
+    }
+  }
+
+  private void checkCaseLabel(CaseLabelTree caseLabel) {
+    var typePattern = getTypePatternFromCaseGroup(caseLabel);
+    typePattern.ifPresent(typePatternTree -> checkTypePatternVariableUsage(typePatternTree.patternVariable()));
+  }
+
+  private static Optional<TypePatternTree> getTypePatternFromCaseGroup(CaseLabelTree caseLabel) {
+    if (caseLabel.expressions().size() == 1
+      && caseLabel.expressions().get(0) instanceof TypePatternTree typePattern
+      && isRecordPattern(typePattern)) {
+      return Optional.of(typePattern);
+    }
+    return Optional.empty();
+  }
+
+  private void checkTypePatternVariableUsage(VariableTree patternVariable) {
+    var secondaryLocationsTrees = new HashSet<Tree>();
+    var type = patternVariable.symbol().type().symbol();
+    var comps = recordComponents(type);
+    for (Tree usage : patternVariable.symbol().usages()) {
+      if (usage.parent() instanceof MemberSelectExpressionTree mse && isNotRecordGetter(mse)) {
+        secondaryLocationsTrees.add(mse);
+      } else {
+        return;
+      }
+    }
+    // only if all the records components are used we report an issue
+    if (secondaryLocationsTrees.size() == comps.size()) {
```

Comment:
Unfortunately this doesn't work quite as one might think. E.g. try the sample:
```
    if (obj instanceof Point p) {
      return p.x() + p.x();
    }
```
It will raise an issue because the size is 2 for both. However, only `x()` is called (2x), never `y()`. Our AST trees don't work well with equals & hashCode, they usually simply inherit `Object`'s implementation. Instead, you probably need to put the identifier names in a set for proper hashing & equality.

Reply:
Argh, that's what I wanted to avoid.. good catch tho!